client/go/dogma/dogma.go

```diff
@@ -0,0 +1,362 @@
+// Copyright 2018 LINE Corporation
+//
+// LINE Corporation licenses this file to you under the Apache License,
+// version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at:
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+/*
+Package dogma provides a Go client for using the Central Dogma API.
+
+Usage:
+
+	import "github.com/line/centraldogma/client/go"
+
+Create a client with the clientID and clientSecret, then use the client to access the
+Central Dogma HTTP APIs. For example:
+
+	clientId := "foo"
+	clientSecret := "bar"
+	client, err := dogma.NewClient("http://localhost:36462", clientId, clientSecret)
+
+	projects, res, err := client.ListProjects(context.Background())
+
+Note that all of the APIs are using the https://godoc.org/context which can pass
+cancellation and deadlines for handling a request.
+*/
+package dogma
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/oauth2/clientcredentials"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+const (
+	DefaultPort = 36462
+
+	defaultScheme   = "http"
+	defaultHostName = "localhost"
+	defaultBaseURL  = defaultScheme + "://" + defaultHostName + ":36462/"
+
+	defaultPathPrefix = "api/v1/"
+	v0PathPrefix      = "api/v0/"
+
+	pathSecurityEnabled = "security_enabled"
+	pathAuthenticate    = v0PathPrefix + "authenticate"
+)
+
+// A Client communicates with the Central Dogma server API.
+type Client struct {
+	client *http.Client // HTTP client which sends the request.
+
+	baseURL *url.URL // Base URL for API requests.
+
+	// Services are used to communicate for the different parts of the Central Dogma server API.
+	project    *projectService
+	repository *repositoryService
+	content    *contentService
+}
+
+type service struct {
+	client *Client
+}
+
+// NewClient returns a Central Dogma client with the specified baseURL, clientID and clientSecret.
+func NewClient(baseURL string, clientID, clientSecret string) (*Client, error) {
```

Comment:
how about 2 separate constructors like

```
NewClient(baseUrl string) *Client, error {
}

NewSecureClient(baseURL string, clientID, clientSecret string) *Client, error {
}
```

And in inner implement we will check if clientID and clientSecret is empty, we will use secure API and opposite?
Not sure about the spec of CD server right now.

Reply:
I think having just one constructor is fine for now so that the client does not have to handle the authentication by itself. Let's add it when we think it really needs :-)